### ISSUE TITLE
How to keep aurelia.json clean

### ISSUE THREAD

**Issue by suamikim** [aurelia/cli#470]
**I'm submitting a feature request**

* **Library Version:**
How can I check the lib version?

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
6.9.4

* **NPM Version:**
4.1.1

* **Browser:**
Request is not browser related

* **Language:**
TypeScript 2.1.5

**Current behavior:**
When installing 3rd party scripts, I have to add them manually to `aurelia_project/aurelia.json`, which will be handled automatically by future versions of the CLI according to some Aurelia announcements.

What happens if I remove a dependency from my project? Will they also be removed automatically by the CLI? If so, is this also true for "sub dependencies" and will there be a check on whether they are required by other dependencies as well?

**What is the motivation / use case for changing the behavior?**

Short example: I've installed package [md5](https://www.npmjs.com/package/MD5) from npm and added the following lines to `aurelia.json`:

```
// ...
"dependencies": [
	// ...
	"md5",
	"crypt",
	"charenc",
	"is-buffer",
	// ...
]
// ...
```

The packages `crypt`, `charenc` & `is-buffer` are currently solely required by `md5`. If I'd like to remove `md5` sometime in the future there are some things I have to consider in order to keep the `aurelia.json` clean and not include unnecessary dependencies in my builds:

* Which dependencies can I remove with `md5`? How do I know which were the sub dependencies of `md5`?
    * Right now I keep them "logically" grouped inside `aurelia.json` but this won't allow me to refactor it in the future (e.g. arrange dependencies alphabetically)...
    * Also since I can't use comments inside JSON there's no way to tell other contributors why I've grouped the dependencies the way I did. They have to know it somehow before touching `aurelia.json`.
* How do I know if any of the sub dependencies I've removed are not needed by some other 3rd party script? Trial & error?

Will those things also handled automatically by the CLI in the future? If not, what are best practices/recommendations for the considerations I've mentioned above?

**Comment by suamikim** [comment]
I'd really appreciate some input on this. How do people who already use the CLI in larger projects actually handle this?

**Comment by AStoker** [comment]
The thing to manage would be your `package.json`. We have a relatively large project, but we don't have too much trouble. We just make sure that if we're modifying the `package.json`, that we also modify the `aurelia.json` to match what we need. If we remove a dependency, then we're removing it from the `package.json` file as well. At that point, it's up to the package manager to decide what to do with it. If you were to run `npm clean`, I believe that's the command that clears things out of the file system that aren't used. And I believe Yarn has a similar feature.  
I'm not sure if this is something that we should manage or not... It sounds like you're more concerned with keeping the file system clean than the `aurelia.json` file. Is that right?

**Comment by suamikim** [comment]
I'm mostly concerned about forgetting to remove sub dependencies from `aurelia.json` and therefore having stuff in my built bundles which is not used at all (see the example with the `md5` package & all its sub dependencies in my initial post).

I don't care about orphaned files on my local disk at all since those can be easily cleaned up using `npm clean` or `npm prune` (as you've mentioned).

Btw. `md5` was one of the first external dependencies that I've started to use in conjunction with Aurelia and so far it turned out to be the only one which actually required explicit entries within `aurelia.json` for its sub dependencies.

In other words: All other libraries I've used so far only required one entry in `aurelia.json` (simple string or config object) which can be simply removed without the "danger" of forgetting its sub dependencies...

So maybe this is not as big of an issue as it seemed to me at first glance...

**Comment by AStoker** [comment]
So I'm not sure if this is something that could be handled by the CLI.  
If I'm understanding you correctly, you have a library, which has dependencies you need to install. When you remove the library, you need to know which dependencies you've installed go along with it. Normally, you would just look in the "dependencies" array of the library you're removing (you _should_ add them there if it is a dependency). 
However, there could be a use case where a dependency of one library is a primary one that you're continuing to use. In that case, when you're trying to clean out dependencies, I'd just do a "find all" in your project for an import matching the dependency in question.  

The reason I don't think the CLI would be able to handle this (I'd love suggestions though if it is possible) is because the CLI can't determine the _intent_ of the originally installed dependency. You add Module A which has a dependency on Module B. So you also install Module B. Turns out that you directly use Module B now too. When you want to remove Module A, the CLI has no idea if you're using Module B or not, so it can't know if it should be removed or not. You could do some fancy regex to determine if it's being imported in your project, but I doubt that's something that would be added to the CLI, and it sounds like it would be better as a gulp task (in which case the user could create their own for their own use cases).

**Comment by suamikim** [comment]
I do understand what you're saying about why the CLI most probably won't be able to resolve all those dependencies & usage information automatically when removing packages and unfortunately don't have an idea on how this could be tackled right now.

However, your post helped me realize that I could change the code from my original post to explicitly state which sub dependencies belong to the 3rd party lib I'm actually using:

`"original" aurelia.json`
```
// ...
"dependencies": [
	// ...
	"md5",
	"crypt",
	"charenc",
	"is-buffer",
	// ...
]
// ...
```

There's no explicit information on why `crypt`, `charenc` & `is-buffer` are listed and some co-workers or "future me" might remove it since it's not used anywhere inside the app.

`"new" aurelia.json`
```
// ...
"dependencies": [
	// ...
	"crypt",
	"charenc",
	"is-buffer",
	{
		"name": "md5",
		"path": "../node_modules/md5",
		"main": "md5",
		"deps": [
			"crypt",
			"charenc",
			"is-buffer"
		]
	}
	// ...
]
// ...
```

This is obviously more verbose but the additional information coming from the dedicated `deps` object helps a lot in my opinion. This way everybody knows, why the packages `crypt`, `charenc` & `is-buffer` are listed within `aurelia.json` even though they are not directly used anywhere within the app code...

Thanks a lot for pointing me in that direction!

**Comment by AStoker** [comment]
Glad it helped you out!
